src/cmd/ksh93/sh/name.c

```diff
@@ -889,16 +889,18 @@ Namval_t *nv_create(const char *name, Dt_t *root, int flags, Namfun_t *dp) {
                         flags &= ~(noscope ? 0 : NV_NOSCOPE);
                     } else if (c) {
                         char *xp = NULL;
-                        ssize_t xlen;
+                        size_t xlen;
                         c = (cp - sp);
                         // Eliminate namespace name.
                         if (shp->last_table && !nv_type(shp->last_table)) {
                             xp = nv_name(shp->last_table);
                             xlen = strlen(xp);
                         }
                         cp = nv_name(np);
-                        if (xp && strncmp(cp, xp, xlen) && cp[xlen] == '.') cp += xlen + 1;
                         copy = strlen(cp);
+                        if (xp && copy > xlen && !strncmp(cp, xp, xlen) && cp[xlen] == '.') {
```

Comment:
You changed `strncmp(cp, xp, xlen)` to `!strncmp(cp, xp, xlen)`. Why ?

Reply:
Because the intent is clearly to test if `xp` is a prefix, that is equal to the first `xlen` chars, of `cp`. The original code tested if they were not equal.

P.S., I instrumented the code and there isn't a single test that causes the body of that `if()` to be executed regardless of whether it is a string equality or inequality test.